File: Problems/Algorithms/131. Palindrome Partitioning/palindrome_partitioning.py
#!/usr/bin/env python3

from typing import List


class Solution:
    def __init__(self) -> None:
        self.results = []
        self.curr = []

    def partition(self, s: str) -> List[List[str]]:
        self.helper(s, 0)
        return self.results
        
    def helper(self, s: str, start: int):
        if start >= len(s):
            self.results.append(self.curr[:])
            return

        for i in range(start, len(s)):
            if self.is_palindrome(s[start:i+1]):
                self.curr.append(s[start:i+1])
                self.helper(s, i+1)
                self.curr.pop()

    def is_palindrome(self, s: str) -> bool:
        return s == s[::-1]

File: Problems/Algorithms/131. Palindrome Partitioning/PalindromePartitioning.kt
class Solution {
    val results: MutableList<List<String>>
    val curr: MutableList<String>

    init {
        results = mutableListOf()
        curr = mutableListOf()
    }

    fun partition(s: String): List<List<String>> {
        helper(s, 0)
        return results
    }
        
    private fun helper(s: String, start: Int) {
        if (start >= s.length) {
            results.add(curr.toMutableList())
            return
        }

        for (i in start..s.length-1) {
            if (isPalindrome(s.substring(start, i+1))) {
                curr.add(s.substring(start, i+1))
                helper(s, i+1)
                curr.removeAt(curr.size-1)
            }
        }
    }

    private fun isPalindrome(s: String): Boolean {
        return s == s.reversed()
    }
}
